sdk/opendp/smartnoise/_ast/tokens.py

```diff
@@ -129,6 +129,58 @@ def find_nodes(self, type_name, not_child_of=None):
         childnodes = [c.find_nodes(type_name, not_child_of) for c in sqlnodes]
         return nodes + flatten(childnodes)
 
+    def visualize(self, color_types={}, n_trunc=None):
+        """
+        Construct the AST graph of the object
+
+        Args:
+            color_type (optional): A dictionnary which contains the type and
+                the color of the nodes.
+                If no color is provided, the displayed node will be black.
+                Example: {str: 'red', Query: 'green'}
+            n_trunc: for visibility, truncate the expressions.
+                By default, there is no truncation.
+
+        Returns:
+            graphviz Digraph
+
+        """
+        def _label_node(expr, n_trunc):
+            str_expr = str(expr)
+            if n_trunc and len(str_expr) > n_trunc:
+                str_expr = str_expr[:n_trunc] + '...'
+            return f"{type(expr).__name__}: {str_expr}"
+
+        def _color_node(expr, color_types):
+            expr_type = type(expr)
+            if expr_type in color_types.keys():
+                return color_types[expr_type]
+            return 'black'
+
+        def _visit_nodes(node, path_to_node):
+            if not isinstance(node, Sql):
+                return
+            for child_node in node.children():
+                if child_node is not None:
+                    path_to_child_node = f"{path_to_node}.{str(child_node)}_{str(id(child_node))}"
+                    graph.node(
+                        path_to_child_node,
+                        _label_node(child_node, n_trunc),
+                        color=_color_node(child_node, color_types),
+                    )
+                    graph.edge(path_to_node, path_to_child_node)
+                    _visit_nodes(child_node, path_to_child_node)
+
+        from graphviz import Digraph
+        graph = Digraph()
+        graph.node(
+            str(self),
+            _label_node(self, n_trunc),
+            color=_color_node(self, color_types),
+        )
+        _visit_nodes(self, str(self))
+        return graph
```

Comment:
Is there any way to test this? I imagine the graph job is expected to have a certain shape. It would be worth having a minimal one node and maybe a 3 node option given it is recursive.

I am reading through more thoroughly now. The PR looks really simple with the current structure. Thanks for taking the time to make it read so well

Reply:
Sorry, just saw this.  I added 4 unit tests; currently just checking that the various calls return Digraphs, but the visualize lines can be uncommented to inspect the rendering, and everything matches as expected.  Note that the unit tests are causing integration tests to fail, due to graphviz not being present.  I just added graphviz to tests/requirements.txt to fix that